### PR TITLE
Chapter 12: modular version

### PR DESCRIPTION
**Pull request by Peter-VanWeert**

### CODE REVIEW

File: Examples/Modules/Chapter 12/Ex12_17/RandomBoxes.cppm [hunk 1]
```diff
@@ -0,0 +1,26 @@
+export module box.random;
+import box;
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 12/Ex12_17/RandomBoxes.cppm:2]: You don't want to export box?

File: Examples/Modules/Chapter 12/Ex12_17/Truckload.cppm [hunk 2]
```diff
@@ -0,0 +1,28 @@
+export module truckload;
+
+export import :shared_box;
+import :package;
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 12/Ex12_17/Truckload.cppm:4]: Shouldn't this be export import?
Ch11 says "the key is that every interface partition must eventually be exported from the module’s so-called primary module interface file" and truckload:package is an interface partition in package.cppm.

File: Examples/Modules/Chapter 12/Ex12_17/Package.cppm [hunk 3]
```diff
@@ -0,0 +1,21 @@
+module truckload:package;
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 12/Ex12_17/Package.cppm:1]: This doesn't compile.
This is a CPPM file, so this must be "export module".
But, later in Truckload.cppm you do "import :package" which is not allowed for interface partitions, so probably you meant this package.cppm file to be package.cpp and an internal partition?

File: Examples/Modules/Chapter 12/Ex12_17/Ex12_17.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
+// Using a linked list
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 12/Ex12_17/Ex12_17.cpp:1]: I couldn't get this Ex12_17 to compile.
After some workaround for VC++, I'm left with:
```
1>F:\PVW\beginning-cpp20\Examples\Modules\Chapter 12\Ex12_17\Truckload.cpp(9,32): error C2061: syntax error: identifier 'Package'
1>F:\PVW\beginning-cpp20\Examples\Modules\Chapter 12\Ex12_17\Truckload.cpp(9,32): error C2143: syntax error: missing ';' before '{'
1>F:\PVW\beginning-cpp20\Examples\Modules\Chapter 12\Ex12_17\Truckload.cpp(9,38): error C2143: syntax error: missing ';' before '}'
```

Comment (MarcGregoire) [reply Examples/Modules/Chapter 12/Ex12_17/Ex12_17.cpp:1]: Did you get it to compile?

File: Examples/Modules/Chapter 12/Ex12_18/Ex12_18.cpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
+// Using nested classes
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 12/Ex12_18/Ex12_18.cpp:1]: Ex12_18 compiles in VC++ :)
Only the ceil() workaround was needed.

File: Exercises/Modules/Chapter 12/Soln12_02/Integer.cppm [hunk 6]
```diff
@@ -0,0 +1,52 @@
+/***************************************************************\
+ This header shows two distinct ways of allowing an Integer
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 12/Soln12_02/Integer.cppm:2]: "this header" ... technically, it's not a header anymore ;-)

File: Exercises/Modules/Chapter 12/Soln12_02/Integer.cppm [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+  void show() const;
+
+private:
+  int m_value{};
+};
+
+// Option 2: use zero a default parameter value
+/*
+class Integer
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 12/Soln12_02/Integer.cppm:36]: export

File: Exercises/Modules/Chapter 12/Soln12_04/Integer.cppm [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  void show() const;
+
+  friend int compare(const Integer& obj1, const Integer& obj2);  // friend compare function
+
+private:
+  int m_value;
+};
+
+// A non-friend function that implements the same function
+int nonFriendCompare(const Integer& obj1, const Integer& obj2);
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 12/Soln12_04/Integer.cppm:21]: Do you want to export this?
I'm too lazy to check the text ... :-p

File: Exercises/Modules/Chapter 12/Soln12_04/Integer.cpp [hunk 9]
```diff
@@ -0,0 +1,57 @@
+import <iostream>;
+module integer;
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 12/Soln12_04/Integer.cpp:2]: reverse import and module lines.

File: Exercises/Modules/Chapter 12/Soln12_08/Soln12_08.cpp [hunk 10]
```diff
@@ -0,0 +1,101 @@
+// Remove Boxes from a Truckload without searching a second time
```
Comment (MarcGregoire) [inline Exercises/Modules/Chapter 12/Soln12_08/Soln12_08.cpp:1]: There is indeed something wrong with the friend declarations.
error C2248: 'Truckload::Iterator::Iterator': cannot access private member declared in class 'Truckload::Iterator'

Comment (MarcGregoire) [reply Exercises/Modules/Chapter 12/Soln12_08/Soln12_08.cpp:1]: bug in VC++?

**Review by MarcGregoire — CHANGES_REQUESTED** [review]

File: Examples/Modules/Chapter 12/Ex12_17/RandomBoxes.cppm (hunk 1 above)
Comment (Peter-VanWeert) [reply Examples/Modules/Chapter 12/Ex12_17/RandomBoxes.cppm:2]: I could. But I didn't in the book, so let's leave it...

File: Examples/Modules/Chapter 12/Ex12_17/Truckload.cppm (hunk 2 above)
Comment (Peter-VanWeert) [reply Examples/Modules/Chapter 12/Ex12_17/Truckload.cppm:4]: It's not an interface partition, it's an implementation partition.

File: Examples/Modules/Chapter 12/Ex12_17/Package.cppm (hunk 3 above)
Comment (Peter-VanWeert) [reply Examples/Modules/Chapter 12/Ex12_17/Package.cppm:1]: There's no law stating that .cppm has to be a module interface file, and that internal partitions have to be .cpp files. I meant for this to be an implementation (or internal, like Microsoft calls) partition, and it is. And apparently I've put it in a .cppm file. Could've put it in a .cpp file as well, but in the book I went for .cppm for all partitions I believe.

File: Examples/Modules/Chapter 12/Ex12_17/Truckload.cppm (hunk 2 above)
Comment (MarcGregoire) [reply Examples/Modules/Chapter 12/Ex12_17/Truckload.cppm:4]: as it is now, it's an interface partition right, it's in Package.cppm

File: Examples/Modules/Chapter 12/Ex12_17/Package.cppm (hunk 3 above)
Comment (MarcGregoire) [reply Examples/Modules/Chapter 12/Ex12_17/Package.cppm:1]: Could be, but doesn't VC++ for example automatically handle .cppm as interface files?
The annoying thing of course is that modules are not yet 100% compliant ...

File: Examples/Modules/Chapter 12/Ex12_17/Ex12_17.cpp (hunk 4 above)
Comment (Peter-VanWeert) [reply Examples/Modules/Chapter 12/Ex12_17/Ex12_17.cpp:1]: Don't remember. It indeed does not compile with the current preview of Visual Studio, no...

File: Examples/Modules/Chapter 12/Ex12_17/Truckload.cppm (hunk 2 above)
Comment (Peter-VanWeert) [reply Examples/Modules/Chapter 12/Ex12_17/Truckload.cppm:4]: The file extension does not determine what type a file is, it is the way you compile it. Either way, .cpp or .cppm, the default compilation that VC++ chooses will be wrong. But I agree, it probably would make more sense to use .cpp for implementation/internal partitions.

File: Exercises/Modules/Chapter 12/Soln12_08/Soln12_08.cpp (hunk 10 above)
Comment (Peter-VanWeert) [reply Exercises/Modules/Chapter 12/Soln12_08/Soln12_08.cpp:1]: I'd say a bug, yes. Have it with all my Truckloads from this point on, but only in the modular version.

**Pull request merged**